src/operator/onErrorReturn.ts

```diff
@@ -0,0 +1,64 @@
+import { Operator } from '../Operator';
+import { Subscriber } from '../Subscriber';
+import { Observable } from '../Observable';
+import { OuterSubscriber } from '../OuterSubscriber';
+
+/**
+ * Instructs an Observable to emit a particular item when it encounters an error, and then terminate normally.
+ * @param {function} selector a function that takes as arguments `err`, which is the error. Whatever value
+ *  is returned by the `selector` will be emitted as the next value, just before terminating.
+ * @return {Observable} an observable that originates from either the source or the observable returned by the
+ *  onErrorReturn `selector` function.
+ * @method onErrorReturn
+ * @owner Observable
+ */
+export function onErrorReturn<T, R>(selector: (err: any) => R): Observable<R> {
+  const operator = new OnErrorReturnOperator<T, R>(selector);
+  return this.lift(operator);
+}
+
+export interface OnErrorReturnSignature<T> {
+  <R>(selector: (err: any) => R): Observable<R>;
+}
+
+class OnErrorReturnOperator<T, R> implements Operator<T, R> {
+  constructor(private selector: (err: any) => R) {
+  }
+
+  call(subscriber: Subscriber<R>, source: any): any {
+    return source._subscribe(new OnErrorReturnSubscriber(subscriber, this.selector));
+  }
+}
+
+/**
+ * We need this JSDoc comment for affecting ESDoc.
+ * @ignore
+ * @extends {Ignored}
+ */
+class OnErrorReturnSubscriber<T, R> extends OuterSubscriber<T, R> {
+  constructor(destination: Subscriber<any>,
+              private selector: (err: any) => R) {
+    super(destination);
+  }
+
+  // NOTE: overriding `error` instead of `_error` because we don't want
+  // to have this flag this subscriber as `isStopped`.
+  error(err: any) {
+    if (!this.isStopped) {
+      let result: R;
+
+      try {
+        result = this.selector(err);
+      } catch (err) {
+        this.destination.error(err);
+        return;
+      }
+
+      this.unsubscribe();
+      (<any>this.destination).remove(this);
+
+      this.destination.next(result);
+      this.destination.complete();
```

Comment:
maybe as same as other code does, could do 

```js
const destination = this.destination;

destination.xxx....
```

since `destination` is being accessed multiple times.